rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+NUMERICAL_COLUMNS = ['age', 'fare']
+# input list to create functional model.
+model_inputs = []
+# input list to feed linear model.
+linear_inputs = []
+for feature_name in CATEGORICAL_COLUMNS:
+	feature_input = tf.keras.Input(shape=(1,), dtype=tf.string, name=feature_name, sparse=True)
+	vocab_list = sorted(dftrain[feature_name].unique())
+	# Map string values to indices
+	x = tf.keras.layers.CategoryLookup(vocabulary=vocab_list, name=feature_name)(feature_input)
+  x = tf.keras.layers.CategoryEncoding(num_categories=len(vocab_list))(x)
+	linear_inputs.append(x)
+	model_inputs.append(feature_input)
+
+for feature_name in NUMERICAL_COLUMNS:
+	feature_input = tf.keras.Input(shape=(1,), name=feature_name)
+	linear_inputs.append(feature_input)
+	model_inputs.append(feature_input)
+
+linear_model = tf.keras.experimental.LinearModel(units=1)
+linear_logits = linear_model(linear_inputs)
+model = tf.keras.Model(model_inputs, linear_logits)
+
+model.compile('sgd', loss=tf.keras.losses.BinaryCrossEntropy(from_logits=True), metrics=['accuracy'])
+
+dataset = tf.data.Dataset.from_tensor_slices((
+	(tf.to_sparse(dftrain.sex, "Unknown"), tf.to_sparse(dftrain.n_siblings_spouses, -1),
+	tf.to_sparse(dftrain.parch, -1), tf.to_sparse(dftrain['class'], "Unknown"), tf.to_sparse(dftrain.deck, "Unknown"),
+	tf.expand_dims(dftrain.age, axis=1), tf.expand_dims(dftrain.fare, axis=1)),
+	y_train)).batch(bach_size).repeat(n_epochs)
+
+model.fit(dataset)
+```
+
+### Workflow 2
+
+The second example gives an instruction on how to transition from categorical feature columns to the proposed layers. Note that one difference for vocab categorical column is that, instead of providing a pair of mutually exclusive `default_value` and `num_oov_buckets` where `default_value` represents the value to map input to given out-of-vocab value, and `num_oov_buckets` represents value range of [len(vocab), len(vocab)+num_oov_buckets) to map input to from a hashing function given out-of-vocab value. In practice, we believe out-of-vocab values should be mapped to the head, i.e., [0, num_oov_tokens), and in-vocab values should be mapped to [num_oov_tokens, num_oov_tokens+len(vocab)).
+
+1. Categorical vocab list column
+
+Original:
+```python
+fc = tf.feature_column.categorical_feature_column_with_vocabulary_list(
+	   key, vocabulary_list, dtype, default_value, num_oov_buckets)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryLookup(
+            vocabulary=vocabulary_list, num_oov_tokens=num_oov_buckets)
+out = layer(x)
+```
+
+2. categorical vocab file column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_vocab_file(
+       key, vocabulary_file, vocabulary_size, dtype,
+       default_value, num_oov_buckets)
+```
+Proposed:
+```python
+x = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+layer = tf.keras.layers.CategoryLookup(
```

Comment:
Ah I see, so here you allow the user to provide a vocabulary file directly, so the tf.data example may not be necessary.  May still be useful if users have vocab that needs to be munged a bit before reading directly.  But less important.

Reply:
I think this layer is more of a complimentary to that, i.e., tf.data can parse records and generate vocab file, of read vocab file and do other processing and still return string tensors. This layer is taken from that and convert things to indices before it gets to embedding.